New Range commands for MCP and PCF expanders

giig1967g opened the pull request:
I have added the following new commands:


1a) `mcpGpioRange, startPin, endPin, value [,optional bitmask]`
1b) `pcfGpioRange, startPin, endPin, value [,optional bitmask]`

2a) `mcpGpioPattern, startPin, endPin, bitmask pattern [,optional bitmask]`
2b) `pcfGpioPattern, startPin, endPin, bitmask pattern [,optional bitmask]`

3a) `mcpMode, pin, mode`
3b) `pcfMode, pin, mode`

4a) `mcpModeRange, startPin, endPin, mode`
4b)` pcfModeRange, startPin, endPin, mode`

5a) `monitorRange, plugin, startPin, endPin`
5b) `unMonitorRange, plugin, startPin, endPin`

Syntax is the following.

1) Range commands 
** Par1=starting pin
** Par2=ending pin (must be higher of starting pin; and maximum 16 pin per command)
** Par3=write value: if 0 (or 1) then assume 0 (or 1) for all the pins in the range; 
** Par4=mask (optional): if not present assume to operate in all pins;
**            if present is used as a mask (1=update, 0=do not update).
**            if number of bit lower than number of pins, then padded with 0;
**            if number of bit higher than number of pins, then it's truncated.
**
**  examples:
**  `mcpgpioRange,1,8,1`: set pins 1 to 8 to 1
**  `mcpgpioRange,3,12,1`: set pins 3 to 12 to 1
**  `mcpgpioRange,5,17,0`: set pins 5 to 17 to 0
**  `mcpgpioRange,3,12,1,525`
**                 mask = '0100001101'
**                 write pattern after mask = 'x1xxxx11x1' where x indicates that the pin will not be changed
**  `mcpgpioRange,3,12,1,973`
**                 mask = 973 = '1111001101'
**     write pattern after mask = '1111xx11x1' where x indicates that the pin will not be changed

2) Pattern commands (ADVANCED USER COMMAND. Requires knowledge of binary numbers.)
** Par1=starting pin
** Par2=ending pin (must be higher of starting pin; and maximum 16 pin per command)
** Par3=write pattern: it's a write pattern. Write 0 or 1. 
                     Example: use decimal number 15 (in binary is 00001111) to set to 1 pin 1,2,3 and 4 and to set to 0 pins 5,6,7,8 
**                     if number of bit lower than number of pins, then padded with 0;
**                     if number of bit higher than number of pins, then it's truncated.
** Par4=mask (optional): if not present assume to operate in all pins; 
**            if present is used as a mask (1=update, 0=do not update).
**            if number of bit lower than number of pins, then padded with 0;
**            if number of bit higher than number of pins, then it's truncated.
**
**  examples:
**  `mcpgpioPattern,1,8,255`
**     write pattern = '1101' that will be padded as: '0000001101'
**     mask not present, assume mask = '1111111111'
**  `mcpgpioPattern,3,12,13`
**     write pattern = '1101' that will be padded as: '0000001101'
**     mask not present, assume mask = '1111111111'
**  `mcpgpioPattern,3,12,525`
**     write pattern = 525 = '100001101'
**     mask not present, assume mask = '1111111111'
**  `mcpgpioPattern, 3, 12, 525, 973`
**     write pattern = 525 = '100001101'
**     mask = 973 = '1111001101'
**     write pattern after mask = '1000xx11x1' where x indicates that the pin will not be changed

3) Mode commands
 ** Par1: PIN to be set
 ** Par2: MODE to be set:
 **             0 = OUTPUT
 **             1 = INPUT PULLUP or INPUT PULLDOWN (only for GPIO16)
 **             2 = INPUT
 ** example: `pcfMode,1,0` (set pin 1 as output)

4) Mode Range commands
 ** Par1: Starting PIN to be set
 ** Par2: Ending PIN to be set
 ** Par3: MODE to be set:
 **             0 = OUTPUT
 **             1 = INPUT PULLUP or INPUT PULLDOWN (only for GPIO16)
 **             2 = INPUT
 ** example: `pcfModeRange,1,8,1` (set pins from 1 to 8 to Input Pulldown)

5) Monitor Range Commands
** Par1: gpio or pcf or mcp
** Par2: Starting pin
** Par3: Ending pin
** example: `monitorRange, pcf, 1, 8`

giig1967g committed:
wip for new expander commands

mcpAll
mcpMulti
pcfAll
pcfMulti
monitorAll
monitorMulti

giig1967g committed:
Merge remote-tracking branch 'upstream/mega' into new/gpiocommands

giig1967g committed:
wip new mcp commands

giig1967g committed:
WIP

Added commands
- mcpGPIORange,pStart,pEnd,0 or 1,optional: mask
- mcpGPIOPattern,pStart,pEnd,writePattern,optional:mask
- monitorRange

giig1967g committed:
WIP

Added
- unMonitorRange,plugin(g or m or p),startPin,endPin

giig1967g committed:
WIP

adding PCF support

giig1967g committed:
WIP

added:
pcfgpiorange
pcfgpiopattern

giig1967g committed:
WIP

fixing stuff

giig1967g committed:
wip

giig1967g committed:
WIP

alpha version

giig1967g committed:
WIP

giig1967g commented:
It's still a work in progress version.
I still need to optimize ad reuse the code.
But the core is there.
I would appreciate if someone could test this PR in order to fix potential bugs before releasing it.

If needed I can attach a build. Just tell me which versions you need.

giig1967g committed:
small changes

giig1967g committed:
Merge remote-tracking branch 'upstream/mega' into new/gpiocommands

tonhuisman commented:
@giig1967g You might want to check the CI build result ([Details](https://travis-ci.org/github/letscontrolit/ESPEasy/builds/750181032?utm_source=github_status&utm_medium=notification) link below) as all are failing because of (probably) a missing #include (couldn't find it yet from the web interface)

giig1967g committed:
fixing Travis

giig1967g committed:
code optimizations

TD-er committed:
Merge branch 'mega' into new/gpiocommands

giig1967g commented:
Anyone tested it?
All builds are working. Only one is failing with the following error:
`Error: The program size (1046112 bytes) is greater than maximum allowed (1044464 bytes)`

I am attaching a simple spreadsheet with automation to create patterns and masks.
The file does not contain macros, just formulas.
[RangeCommandBuilder.xlsx](https://github.com/letscontrolit/ESPEasy/files/5735499/RangeCommandBuilder.xlsx)

TD-er commented:
I have not tested it and for some builds like VCC_SD_MDNS builds are "normal" to fail due to binary size.

uzi18 commented:
@giig1967g why not just simply extend arguments to support range, like: mcpGpio, 1-5, value

TD-er commented:
Hmm during our discussion about this we did not think about such range specifiers.
Maybe it is more clear?
On the other hand you can't create such a range from variables, for example if you try to loop over GPIOs to create a light effect.

uzi18 commented:
Yes, you need to get 2nd variable as string and check if there is special char like -
@TD-er is it impossible to run something like this in rules:
 mcpGpio, %v1-%v5, 1

tonhuisman commented:
> is it impossible to run something like this in rules:
> mcpGpio, %v1-%v5, 1

That will probably subtract values, I'd expect `mcpGpio %v1-v5%,1` should be possible to implement, but we explicitly prohibited dashes in task and variable names as they are processed as operators...

uzi18 commented:
These are user variables

tonhuisman commented:
> These are user variables

I was already anticipating on a generic implementation, but `[dummy#var1-dummy#var2]` will cause a whole lot of other issues during parsing, so this kind of usage would likely require implementation of proper string-argument handling within ESPEasy, and I don't see that happen soon. (As in low-prio?)

giig1967g commented:
> @giig1967g why not just simply extend arguments to support range, like: mcpGpio, 1-5, value

Well, we thought it was more friendly to add specific commands to avoid confusion
Of course other solutions could have been possible...

giig1967g commented:
> @giig1967g why not just simply extend arguments to support range, like: mcpGpio, 1-5, value

Well, we thought it was more friendly to add specific commands to avoid confusion
Of course other solutions could have been possible...

TD-er commented:
If you need to create a more intuitive range, then I suggest to use `..` or `...`
Using a - is also parsed as part of a formula (not entirely sure if you need to prepend it with a `=`) so it will be rather confusing, at least for the parser.

Another problem is when considering parameter positions.
If you have a command like this:
```
cmd,start,end,otherparam
```
What will then be the second parameter if you change it to something like this:
```
cmd,start...end,otherparam
```

Another option is to introduce a range specifier to be used anywhere in the software, like this:
`[1-5]` like it is done in a regular expression.

 But I'm afraid this may lead to a "we can do it all, which may take forever", which is something I would like to avoid if possible ;)
Maybe after 3 years it is time not to open another Pandora's box and work towards a release label

uzi18 commented:
Have started only idea for you, it could be also 2 or 3 dots

uzi18 commented:
With such projects it is always never ending story

TD-er commented:
> With such projects it is always never ending story

Yep, so you won't get rid of me for a (long) while :)
All I wanted to say is, great ideas, but let's not implement it all at once or else this will be another PR running for years.

giig1967g committed:
Merge remote-tracking branch 'upstream/mega' into new/gpiocommands

giig1967g committed:
Moved PLUGIN_REQUEST to core

1) Fixed some code

2) Moved to core:
- [plugin#gpio#pinstate#x] or [plugin#g#pinstate#x]
- [plugin#pcfgpio#pinstate#x] or [plugin#p#pinstate#x]
- [plugin#mcpgpio#pinstate#x] or [plugin#m#pinstate#x]

3) Added:
- [plugin#mcpgpio#pinrange#x-y] that can be shortened to [plugin#m#pinrange#x-y]
- [plugin#pcfgpio#pinrange#x-y] that can be shortened to [plugin#p#pinrange#x-y]
returns the pin states of the selected range

TD-er committed:
Merge branch 'mega' into new/gpiocommands

TD-er committed:
Fix merge issues SendStatusOnlyIfNeeded

SendStatusOnlyIfNeeded now needs to have the event, from which it takes the event source.
This function signature change had to be fixed after the merge from the mega branch.

giig1967g committed:
Update GPIO.cpp

small cosmetic fix

giig1967g committed:
Merge branch 'new/gpiocommands' of https://github.com/giig1967g/ESPEasy into new/gpiocommands

giig1967g committed:
Merge remote-tracking branch 'upstream/mega' into new/gpiocommands

giig1967g committed:
added documentation

tonhuisman commented on docs/source/Plugin/P001_commands_GPIO.repl:
```diff
@@ -117,18 +117,26 @@ Supported hardware: |P000_usedby_GPIO|
   "
   "
-  ``Monitor,<GPIO>``
+  ``Monitor,G,<GPIO>``
```
What is that 'G' supposed to do?

giig1967g replied:
the command syntax is:
monitor,gpio,pin
monitor,pcf,pin
monitor,mcp,pin

that can be shrinked to:
monitor,g,pin
monitor,p,pin
monitor,m,pin

giig1967g committed:
Added documentation for [PLUGIN#MCP#PINRANGE#x-y]

Added documentation for [PLUGIN#MCP#PINRANGE#x-y] and [PLUGIN#PCF#PINRANGE#x-y]

TD-er commented on docs/source/Rules/Rules.rst:
```diff
@@ -474,6 +474,16 @@ You must not use the task names ``Plugin``, ``var`` ``int`` as these have specia
+
+For expanders you can use also the following:
+
+``[Plugin#MPC#PinRange#x-y]`` to get the pin state of a range of MCP pins from x o y.
```
MCP I presume :)

giig1967g replied:
yeah right!

giig1967g committed:
fixed typo

TD-er committed:
Merge branch 'mega' into new/gpiocommands

TD-er committed:
[Docs] Fix some documentation inconsistencies

TD-er commented:
I fixed some documentation inconsistencies.
For example the command overview had lost its plugin headers with commands and the list of GPIO commands also needed some sub headers to separate the internal/MCP/PCF commands.

I will make a test build for this.
After the next nightly build I will merge this so we have a clear separation between builds to make sure any new reported issue either is caused by this PR or something else.

giig1967g commented:
> I will make a test build for this.
> After the next nightly build I will merge this so we have a clear separation between builds to make sure any new reported issue either is caused by this PR or something else.

Agreed

TD-er commented:
As promised :)
